src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -199,28 +199,33 @@ private void UpdateSize()
                 if (extraField.Size < sizeof(long))
                     return true;
 
-                long value64 = reader.ReadInt64();
+                // Advancing the stream (by reading from it) is possible only when:
+                // 1. There is an explicit ask to do that (valid files, corresponding boolean flag(s) set to true, #77159).
+                // 2. When the size indicates that all the information is available ("slightly invalid files", #49580).
+                bool readAllFields = extraField.Size >= sizeof(long) + sizeof(long) + sizeof(long) + sizeof(int);
+
+                long value64 = readUncompressedSize || readAllFields ? reader.ReadInt64() : -1;
                 if (readUncompressedSize)
                     zip64Block._uncompressedSize = value64;
 
                 if (ms.Position > extraField.Size - sizeof(long))
                     return true;
 
-                value64 = reader.ReadInt64();
+                value64 = readCompressedSize || readAllFields ? reader.ReadInt64() : -1;
                 if (readCompressedSize)
                     zip64Block._compressedSize = value64;
 
                 if (ms.Position > extraField.Size - sizeof(long))
                     return true;
```

Comment:
Should remaining fields be set to -1 if we return before attempting to read them?

Reply:
If the idea is to keep this PR minimal for backporting, this should probably be done later.

An alternative for this could be to set all those values to `-1` initially, then only the ones that are read before returning early will acquire a different value than -1.